File: src/clients/awb_esp32_client/lib/common/ByteArrayConverter.cpp
#include "ByteArrayConverter.h"

using byte = unsigned char;

// static function to convert unsigned int to an array of 4 byte
byte *ByteArrayConverter::UintTo4Bytes(unsigned int value)
{
    return new byte[4]{(byte)(value >> 24), (byte)(value >> 16), (byte)(value >> 8), (byte)value};
}

// static function to convert an array of 4 byte to unsigned int
unsigned int ByteArrayConverter::UintFrom4Bytes(byte value[])
{
    return (unsigned int)(value[0] << 24 | value[1] << 16 | value[2] << 8 | value[3]);
}

// to prevent usage of the reserved bytes for packet headers, we split the bytes into 2 parts
byte *ByteArrayConverter::SplitByte(byte value)
{
    auto result = new byte[2];
    if (value > 200)
    {
        result[0] = 200;
        result[1] = (byte)(value - 200);
    }
    else
    {
        result[0] = value;
        result[1] = 0;
    }
    return result;
}

// to prevent usage of the reserved bytes for packet headers, we split the bytes into 2 parts
byte *ByteArrayConverter::SplitBytes4(byte value[4])
{
    auto result = new byte[8];
    for (int i = 0; i < 4; i++)
    {
        if (value[i] > 200)
        {
            result[i * 2] = 200;
            result[i * 2 + 1] = (byte)(value[i] - 200);
        }
        else
        {
            result[i * 2] = value[i];
            result[i * 2 + 1] = 0;
        }
    }
    return result;
}

byte *ByteArrayConverter::UnSplitBytes8(byte value[8])
{
    auto result = new byte[4];
    for (int i = 0; i < 4; i++)
        result[i] = (byte)(value[i * 2] + value[i * 2 + 1]);
    return result;
}

byte ByteArrayConverter::UnSplitBytes2(byte value[2])
{
    return (byte)(value[0] + value[1]);
}

File: src/desktop/PacketLogistic/PacketLogistics/ComPorts/ByteArrayConverter.cs
// Send and receivce data to/from ESP-32 microcontroller
// (C) 2025 Daniel Springwald      -     Bochum, Germany
// https://daniel.springwald.de - segfault@springwald.de
// All rights reserved    -   Licensed under MIT License

using System.Runtime.CompilerServices;
using System.Text;

[assembly: InternalsVisibleTo("PacketLogisticsTests")]
namespace PacketLogistics.ComPorts
{
    public static class ByteArrayConverter
    {
        const byte splitPoint = 128;

        /// <summary>
        /// to prevent usage of the reserved bytes for packet headers, we split the bytes into 2 parts
        /// </summary>
        public static IEnumerable<byte> SplitByte(byte b) => SplitBytes(new byte[] { b });

        /// <summary>
        /// to prevent usage of the reserved bytes for packet headers, we split the bytes into 2 parts
        /// </summary>
        public static IEnumerable<byte> SplitBytes(byte[] bytes)
        {
            foreach (var b in bytes)
            {
                if (b > splitPoint)
                {
                    yield return splitPoint;
                    yield return (byte)(b - splitPoint);
                }
                else
                {
                    yield return b;
                    yield return 0;
                }
            }
        }

        public static IEnumerable<byte> UnSplitBytes(byte[] bytes)
        {
            for (var i = 0; i < bytes.Length - 1; i += 2)
                yield return (byte)(bytes[i] + bytes[i + 1]);
        }

        public static byte[] UintTo4Bytes(uint value)
        {
            return new byte[4] { (byte)(value >> 24), (byte)(value >> 16), (byte)(value >> 8), (byte)value };
        }

        public static uint UintFrom4Bytes(byte[] value)
        {
            if (value.Length != 4) throw new ArgumentOutOfRangeException(paramName: nameof(value), message: "value must be 4 bytes long");
            return (uint)(value[0] << 24 | value[1] << 16 | value[2] << 8 | value[3]);
        }

        public static byte[] AsciiStringToBytes(string value)
        {
            if (string.IsNullOrWhiteSpace(value)) return Array.Empty<byte>();
            return Encoding.ASCII.GetBytes(value);
        }

        public static string BytesToAsciiString(byte[] value)
        {
            if (value == null) return string.Empty;
            return Encoding.ASCII.GetString(value);
        }

        public static byte[]? GetNextBytes(byte[] value, int count, ref int pos)
        {
            if (pos + count > value.Length) return null;
            var result = value[pos..(pos + count)];
            pos += count;
            return result;
        }

        public static bool AreEqual(byte[]? value1, byte[]? value2)
        {
            if (value1 == null && value2 == null) return true;
            if (value1 == null || value2 == null) return false;
            if (value1.Length != value2.Length) return false;
            for (int i = 0; i < value1.Length; i++)
            {
                if (value1[i] != value2[i]) return false;
            }
            return true;

        }
    }
}
